### PR TITLE
[5.2] Impliment nested wheres in has queries

### PR DESCRIPTION
**Pull request by wleona3**
Implemented based on https://github.com/laravel/framework/pull/13785. Existing scope methods worked, so they've been renamed and aliased.

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 1]
```diff
@@ -1208,26 +1216,50 @@ protected function applyScope($scope, $builder)
...
+
+    /**
+     * Determine if the scope added after the given offset should be nested.
+     *
+     * @param  \Illuminate\Database\Query\Builder  $query
+     * @param  int  $originalWhereCount
+     * @return bool
+     */
+    protected function shouldNestWheres(QueryBuilder $query, $originalWhereCount)
+    {
```
Comment (acasar) [inline src/Illuminate/Database/Eloquent/Builder.php]: @wleona3 I don't think we really need to alias ``shouldNestWheresForScope``. Just use the existing method.

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 2]
```diff
@@ -1208,26 +1216,50 @@ protected function applyScope($scope, $builder)
...
+
+    /**
+     * Nest where conditions by slicing them at the given where count.
+     *
+     * @param  \Illuminate\Database\Query\Builder  $query
+     * @param  int|array  $whereCounts
+     * @return void
+     */
+    protected function nestWheres(QueryBuilder $query, $whereCounts)
+    {
```
Comment (acasar) [inline src/Illuminate/Database/Eloquent/Builder.php]: @wleona3 same here.

File: tests/Database/DatabaseEloquentBuilderTest.php [hunk 3]
```diff
@@ -526,6 +526,21 @@ public function testHasWithContraintsAndHavingInSubquery()
...
+    {
+        $model = new EloquentBuilderTestModelParentStub;
+
+        $builder = $model->where('bar', 'baz');
+        $builder->whereHas('foo', function ($q) {
+            $q->having('bam', '>', 'qux');
+            $q->where('baam', '>', 'quux');
+            $q->orWhere('baaam', '>', 'quuux');
+        })->where('quux', 'quuuux');
+
```
Comment (acasar) [inline tests/Database/DatabaseEloquentBuilderTest.php]: @wleona3 Can you maybe use some more "human readable" example? like users, posts, etc. instead of quuux and baaam...

**Comment by acasar** [comment]
@wleona3 All in all, I was thinking we could maybe extract the shared logic between ``callScope`` and ``has`` methods.

I'd suggest something like this: https://github.com/acasar/framework/commit/a21028f1efd72ff98b27919dd9ec52621bfbaa06

**Comment by wleona3** [comment]
Agreed and updated.

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 4]
```diff
@@ -1239,7 +1254,7 @@ protected function nestWheresForScope(QueryBuilder $query, $whereCounts)
         // collection (0 and total where count) while also flattening the array
         // and extracting unique values, ensuring that all wheres are sliced.
         $whereOffsets = collect([0, $whereCounts, count($allWheres)])
-                    ->flatten()->unique();
+            ->flatten()->unique();
```
Comment (acasar) [inline src/Illuminate/Database/Eloquent/Builder.php]: Please revert this line to original formatting.

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 5]
```diff
@@ -1220,7 +1235,7 @@ protected function shouldNestWheresForScope(QueryBuilder $query, $originalWhereC
     }
 
     /**
-     * Nest where conditions by slicing them at the given where count.
+     * Alias fpr the nestWheres method.
```
Comment (acasar) [inline src/Illuminate/Database/Eloquent/Builder.php]: Please revert this line to original formatting.

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 6]
```diff
@@ -1208,7 +1223,7 @@ protected function applyScope($scope, $builder)
     }
 
     /**
-     * Determine if the scope added after the given offset should be nested.
+     * Alias for the shouldNestWheres method.
```
Comment (acasar) [inline src/Illuminate/Database/Eloquent/Builder.php]: Please revert this line to original wording.

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 7]
```diff
@@ -1137,14 +1137,29 @@ protected function callScope($scope, $parameters)
...
+     * @param  callable $callback
+     * @param  array $parameters
+     * @param  \Illuminate\Database\Query\Builder $query
+     * @return mixed
+     */
+    protected function applyCallback(callable $callback, $parameters = [], $query = null)
+    {
+        if (! $query) {
+            $query = $this->getQuery();
+        }
```
Comment (acasar) [inline src/Illuminate/Database/Eloquent/Builder.php]: Maybe a one-liner here? ``$query = $query ?: $this->getQuery();``

**Comment by acasar** [comment]
Looking better now :) Can you apply the formatting changes and squash the commits?

**Comment by wleona3** [comment]
Changes made and squashed

File: src/Illuminate/Database/Eloquent/Builder.php (hunk 4 above)
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Builder.php]: Yes, please revert this change.

**Comment by GrahamCampbell** [comment]
One change remaining to be reverted please.

**Comment by wleona3** [comment]
Whoops. Corrected.

**Comment by GrahamCampbell** [comment]
Thanks. :)

**Comment by acasar** [comment]
@taylorotwell this is now ready for review.

Basically, ``orWhere`` constraints in ``whereHas`` are now automatically nested, same as with local and global scopes. Example:

```php
Model::whereHas('relation', function ($query) {
    $query->where('foo', 'bar')->orWhere('foo', 'baz');
});
```

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 8]
```diff
@@ -1137,14 +1137,27 @@ protected function callScope($scope, $parameters)
...
+
+    /**
+     * Apply the given callback to a supplied (or the current) builder instance.
+     *
+     * @param  callable $callback
+     * @param  array $parameters
+     * @param  \Illuminate\Database\Query\Builder $query
+     * @return mixed
+     */
+    protected function applyCallback(callable $callback, $parameters = [], $query = null)
```
Comment (taylorotwell) [inline src/Illuminate/Database/Eloquent/Builder.php]: What is the purpose of this method? Could it have a better more descriptive name than `applyCallback` which sounds very vague and general?

Comment (acasar) [inline src/Illuminate/Database/Eloquent/Builder.php]: It applies the callback in a way that doesn't cause any issues logical issues with boolean order.

To put it simply - "or where" conditions are nested.

Comment (wleona3) [inline src/Illuminate/Database/Eloquent/Builder.php]: How about something like applyCallbackAndNestWheres?

**Comment by theTrip73** [comment]
Can this be pulled into 5.1?

**Comment by GrahamCampbell** [comment]
> Can this be pulled into 5.1?

Sorry, no. Our LTS branch gets big fixes only.

**Comment by theTrip73** [comment]
You're saying that this -- https://github.com/laravel/framework/pull/13785#issuecomment-222558695 -- isn't considered a bug?

**Comment by GrahamCampbell** [comment]
No. It's a feature.

**Comment by theTrip73** [comment]
orWhere() not nesting properly inside of a whereHas() is a feature? How so?

**Comment by GrahamCampbell** [comment]
> You can have joins and other clauses in whereHas so we can't nest everything inside where

**Comment by GrahamCampbell** [comment]
The ability to do that is not **required**, it's a feature.

**Comment by taylorotwell** [comment]
No idea what Graham is talking about in regards to "being a feature", so apologize for that... but can you not do this?

->where(function ($q) {
$q->where('foo')->orWhere('bar')
});

**Comment by GrahamCampbell** [comment]
@taylorotwell The point is you can achieve the same thing without nesting. The ability to nest these is a feature.

**Comment by theTrip73** [comment]
@taylorotwell, I understand what @GrahamCampbell is saying. I see both sides of this discussion. Intuitively, it is weird to have to force the nesting with an extra ->where() closure, but I also understand why auto-nesting it causes other problems in the query builder.

All said, the extra ->where() closure isn't the worst solution.

Thanks for the explanations and suggestions.
